compiler/rustc_typeck/src/check/cast.rs

```diff
@@ -201,8 +205,76 @@ fn make_invalid_casting_error<'a, 'tcx>(
     )
 }
 
+pub enum CastCheckResult<'tcx> {
+    Ok,
+    Deferred(CastCheck<'tcx>),
+    Err(ErrorGuaranteed),
+}
+
+pub fn check_cast<'tcx>(
+    fcx: &FnCtxt<'_, 'tcx>,
+    expr: &'tcx hir::Expr<'tcx>,
+    expr_ty: Ty<'tcx>,
+    cast_ty: Ty<'tcx>,
+    cast_span: Span,
+    span: Span,
+) -> CastCheckResult<'tcx> {
+    if cast_ty.is_dyn_star() {
+        check_dyn_star_cast(fcx, expr, expr_ty, cast_ty)
```

Comment:
Can you explain why this is needed? Why can't this live in `CastCheck`, that is?

Reply:
When a dyn* cast check succeeds, there aren't any deferred checks we need to worry about, so it doesn't really make sense to build a `CastCheck`. 

I guess another way to do this would be to make `CastCheck::new` return `Result<Option<CastCheck>, ErrorGuaranteed>`, but this way seemed cleaner to me. I'm happy to change it though if you think that's better.